library/kani/src/slice.rs

```diff
@@ -60,20 +60,9 @@ impl<T, const MAX_SLICE_LENGTH: usize> AnySlice<T, MAX_SLICE_LENGTH> {
     {
         let any_slice = AnySlice::<T, MAX_SLICE_LENGTH>::alloc_slice();
         unsafe {
-            let mut i = 0;
-            // Note: even though the guard `i < MAX_SLICE_LENGTH` is redundant
-            // since the assumption above guarantees that `slice_len` <=
-            // `MAX_SLICE_LENGTH`, without it, CBMC fails to infer the required
-            // unwind value, and requires specifying one, which is inconvenient.
-            // CBMC also fails to infer the unwinding if the loop is simply
-            // written as:
-            //     for i in 0..slice_len {
-            //         *(ptr as *mut T).add(i) = any();
-            //     }
-            while i < any_slice.slice_len && i < MAX_SLICE_LENGTH {
-                *any_slice.ptr.add(i) = any();
-                i += 1;
-            }
+            let i = any();
+            assume(i < any_slice.slice_len);
+            *any_slice.ptr.add(i) = any();
```

Comment:
Wouldn't that leave the entries uninitialized (i.e. poison) as opposed to initialized with a non-deterministic value?

Reply:
Yes, but do we care about the entire content of the slice at his point or we want to make sure that if we use this content in an operation later, we will consider the case where `any()` part of the slice could have a non-deterministic value?

Instead of using a loop, we can create a non-deterministic index, make sure (with an assumption) that the index is within bounds, and then non-deterministically select a section of slice and assign it to `any()`. During symbolic execution, whenever we access this slice, it will consider the path where the section is assign to `any()`. Thus, we are basically taking advantage of the symbolism in symbolic execution instead of unrolling the loop during GOTO transformation.

For instance,
```rust
//consider p = poison value;
slice := [p, p, …, p, …, p]

i = any()
assume( i < slice.len())

slice[i] = any()
slice:= [p,p, …, p, any(), p, …, p, p] // any() could be placed in any() location up to slice.len()

// now let's assert something about slice
j = any();
assert(slice[j] == p); // any assertion on the content of slice will fail, since any location could be equal to any()
```

Reply:
@celinval does that clarify your question too? ☝️

Reply:
@feliperodri Reading uninitialized memory in Rust is undefined behavior: https://doc.rust-lang.org/reference/behavior-considered-undefined.html

The compiler may do anything with those values unless we initialize them.

Reply:
I suspect the real solution is to move `any_slice` from a rust implementation to a kani intrinsic, like we do for `any`.

Reply:
Oh! So we should not implement this here in Rust, maybe at the GOTO level only. Is there a way where I only include the body of this function later during codegen?

Reply:
We use hooks for implementing functions in codegen. For example, `kani::any_raw_inner` is handled here: https://github.com/model-checking/kani/blob/51b758b246198b47dcb3ed3c485aa47ec9398ba1/kani-compiler/src/codegen_cprover_gotoc/overrides/hooks.rs#L168

Out of curiosity, have you run into a case where `any_slice` in its current form is causing a performance issue?